### PR TITLE
Document frontmatter can now be surrounded by ---

### PR DESCRIPTION
**Pull request by freiguy1**
#431 
Documents can optionally start with a `---` now.  The code I added essentially checks if the document has front matter surrounded by `---` and if so, it removes the first line and then the logic for separating the front matter stays the same. I added a couple tests as well.

Questions and suggestions are welcome and encouraged! I tried a few ideas before settling on this.

### CODE REVIEW

File: src/cobalt_model/document.rs [hunk 1]
```diff
@@ -74,6 +74,8 @@ fn deprecated_split_front_matter(content: &str) -> Result<(Option<&str>, &str)>
         static ref FRONT_MATTER_DIVIDE: regex::Regex = regex::Regex::new(r"(\A|\n)---\s*\r?\n").unwrap();
     }
     if FRONT_MATTER_DIVIDE.is_match(content) {
+        warn!("Front matter should have a separator above and below. Using just a trailing separator is deprecated.");
```
Comment (epage) [inline src/cobalt_model/document.rs]: Recommendation: "Trailing separators are deprecated. We recommend frontmatters be surrounded, above and below, with `---`".

**Comment by freiguy1** [comment]
How & where would I do this?
> Could you call out that this is a breaking change for anyone who had an empty frontmatter that later had --- showing up in their document?

**Comment by epage** [comment]
> How & where would I do this?

Sorry I didn't specify.  I'd recommend putting it in the commit message.  My preferred style so I'm less likely to miss it is [conventional](https://github.com/conventional-changelog/conventional-changelog/blob/a5505865ff3dd710cf757f50530e73ef0ca641da/conventions/angular.md)

**Comment by epage** [comment]
@freiguy1 I've not seen an update in a while (and that is fine if you are busy).  I just want to check in if there is anything you'd like help with, including completing this if you can't get to it.  Whatever works for you; I just don't want to block you using this, frustrate you with steps to get in, etc.

**Comment by freiguy1** [comment]
Thanks for being so accommodating! You all have been a pleasure to work with as I'm still unfamiliar with this process. When you say

> Could you change your commit style to match conventional? It'll help remind me to include this in the changelog.

Does that mean I can change the commit messages for previous commits or just going forward? I didn't know it was possible to change previous commit messages.

**Comment by freiguy1** [comment]
I must admit I got a little lost while trying to change the commit message on my first commit - so I added the BREAKING CHANGE indicator to the most recent commit (which just had some code formatting fixes).

**Comment by epage** [comment]
> I must admit I got a little lost while trying to change the commit message on my first commit - so I added the BREAKING CHANGE indicator to the most recent commit (which just had some code formatting fixes).

its ok, I'll see about cleaning it up.

**Comment by epage** [comment]
Thanks for doing all that work and being patient with my suggestions!

**Comment by berkus** [comment]
🥇

**Pull request merged**